Better Configuration in TypeScript with the `satisfies` Operator/thompson.md

```diff
@@ -0,0 +1,250 @@
+
+```js
+// 🚩 타입스크립트 컴파일은 되지만, `routes.NONSENSE`는 실제로 존재하지 않음.
+routes.NONSENSE.path;
+```
+
+왜 그럴까요? 이것은 우리의 `Routes` 타입이 키로 어떤 문자열이든 받을 수 있기 때문입니다. 그렇기 때문에 타입스크립트는 간단한 오타부터 말이 안되는 키를 포함한 모든 종류의 키 접근을 허용합니다. 또한 우리는 우리의 IDE로부터 유효한 route 키를 자동완성 해주는 도움을 받을 수 없습니다.  
+당신은 "그럼 `as` 키워드는 어때요?"라고 말할 수 있습니다. 그러면 그것도 간단하게 살펴볼 수 있습니다.
+
+```ts
+type Route = { path: string; children?: Routes };
+type Routes = Record<string, Route>;
+
+const routes = {
+  AUTH: {
+    path: "/auth",
+  },
+} as Routes;
+```
+
+이것은 타입스크립트에서 일반적인 관행이지만, 사실 상당히 위험합니다.
+
+위에서와 같은 문제가 있을 뿐만 아니라, 존재하지 않는 키값을 사용할 수 있고 TypeScript will look the other way(여기를 어떻게 번역해야 할 지 모르겠어요)
+
+```ts
+type Route = { path: string; children?: Routes };
+type Routes = Record<string, Route>;
+
+const routes = {
+  AUTH: {
+    path: "/auth",
+    // 🚩 타입스크립트 컴파일은 제대로 되지만, 이것은 유효한 속성이 아닙니다!
+    nonsense: true,
+  },
+} as Routes;
+```
+
+넓게 말하면, 보통의 경우 타입스크립트에서 `as` 키워드를 사용하는 것을 피하세요.
+
+## `Satisfies`가 목숨을 살린다
+
+이제, 우리는 마침내 위의 예시를 새로운 `satisfies` 키워드로 재작성할 수 있습니다.
+
+```ts
+type Route = { path: string; children?: Routes };
+type Routes = Record<string, Route>;
+
+const routes = {
+  AUTH: {
+    path: "/auth",
+  },
+} satisfies Routes;
+```
+
+이를 통해, 우리는 우리가 바라던 모든 적절한 타입 체크를 할 수 있습니다.
+
+```ts
+routes.AUTH.path; // ✅
+routes.AUTH.children; // ❌ routes.auth에는 `children` 속성이 없습니다.
+routes.NONSENSE.path; // ❌ routes.NONSENSE는 존재하지 않습니다.
+```
+
+그리고 물론 IDE 자동완성도 됩니다.
+![자동완성](https://cdn.builder.io/api/v1/image/assets%2FYJIGb4i01jvw0SRdL5Bt%2F4cd04efbeb764f6cb29bfce5cf091be9?format=webp&width=2000)
+
+우리는 아래와 같은 좀 더 복잡한 예제도 다룰 수 있습니다.
+
+```ts
+type Route = { path: string; children?: Routes };
+type Routes = Record<string, Route>;
+
+const routes = {
+  AUTH: {
+    path: "/auth",
+    children: {
+      LOGIN: {
+        path: "/login",
+      },
+    },
+  },
+  HOME: {
+    path: "/",
+  },
+} satisfies Routes;
+```
+
+이전에 사용했던 좀 더 일반적인 방식과 반대로, 정확히 우리가 구성한 대로 IDE의 자동완성과 모든 객체 트리에 대한 타입 체크를 할 수 있습니다.
+![동영상](./thompson-video.gif)
+
+```ts
+routes.AUTH.path; // ✅
+routes.AUTH.children.LOGIN.path; // ✅
+routes.HOME.children.LOGIN.path; // ❌ routes.HOME에 `children` 속성이 없습니다.
+```
+
+완벽하게 우리가 원하던 것입니다.
+
+## `as const`와 결합
+
+당신이 마주할 수 있는 마지막 상황은, 평범하게 `satisfies`를 사용했을 때, 우리가 구성한 객체는 생각했던 것보다 더 느슨해질 수 있습니다. (is captured를 뭐라고 번역해야 할까요?)  
+예를 들어, 아래 코드에서
+
+```ts
+const routes = {
+  HOME: { path: "/" },
+} satisfies Routes;
+```
+
+만약 우리가 `path` 속성의 타입을 체크하면, 우리는 `string` 타입을 얻습니다.
+
+```ts
+routes.HOME.path; // Type: string
+```
+
+그러나 구성과 관해서는, [const 단언](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) (`as const`)가 빛을 발합니다. 만약 여기서 `as const`를 사용한다면 우리는 더 정확한 문자열 리터럴 `'/'` 타입을 얻을 수 있습니다.
+
+```ts
+const routes = {
+  HOME: { path: "/" },
+} as const;
+
+routes.HOME.path; // Type: '/'
+```
+
+"그래, 이건 교묘한 수법인데 내가 왜 신경 써야돼"와 같이 생각할 수 있습니다. 그리나 아래와 같이 매개변수로 받는 정확한 경로까지 type-safe한 메소드를 고려해봅시다.(여기 문장을 어떻게 번역해야할 지 애매함)
```

Comment:
> This may seem like “ok, that’s a neat trick but why do I care”

이 부분은 저도 비슷하게 되었습니다.

> but let’s consider if we had a method like this that was type-safe all the way down to the exact paths it accepts:

이 부분은 저는 다음과 같이 조금 단어들을 빼서 해석하였습니다.
"정확한 경로까지 type-safe 하게 사용할 수 있는 방법인지 고려해봐야 합니다." 

저도 경로까지 type-safe하게 하는 방법 쪽으로 해석하면 좀 더 자연스럽지 않을까 합니다. ㅎㅎㅎㅎ

(그리나 -> 그러나 )ㅎㅎㅎㅎㅎ